core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java

```diff
@@ -544,19 +558,19 @@ public ChannelPipeline getPipeline() throws Exception {
                 requestDecoder.setMaxCumulationBufferComponents(transport.maxCompositeBufferComponents);
             }
             pipeline.addLast("decoder", requestDecoder);
-            pipeline.addLast("decoder_compress", new ESHttpContentDecompressor(transport.compression));
+            pipeline.addLast("decoder_compress", new HttpContentDecompressor());
             HttpChunkAggregator httpChunkAggregator = new HttpChunkAggregator((int) transport.maxContentLength.bytes());
             if (transport.maxCompositeBufferComponents != -1) {
                 httpChunkAggregator.setMaxCumulationBufferComponents(transport.maxCompositeBufferComponents);
             }
             pipeline.addLast("aggregator", httpChunkAggregator);
-            if (SETTING_CORS_ENABLED.get(transport.settings())) {
-                pipeline.addLast("cors", new CorsHandler(transport.getCorsConfig()));
-            }
             pipeline.addLast("encoder", new ESHttpResponseEncoder());
             if (transport.compression) {
                 pipeline.addLast("encoder_compress", new HttpContentCompressor(transport.compressionLevel));
             }
+            if (SETTING_CORS_ENABLED.get(transport.settings())) {
```

Comment:
@abeyad Sorry that I did not state it clearly enough. The header itself would still be available but it is read by `HttpContentEncoder#messageReceived()` was not called if the CORS handler came before in the pipeline. When the response is about to be sent, Netty calls `HttpContentEncoder#messageReceived()` which relies on a field being set by `messageReceived()`. That's what caused the problem.